addon/Feeds/TopSitesFeed.js

```diff
@@ -10,26 +10,60 @@ const getScreenshot = require("addon/lib/getScreenshot");
 const {isRootDomain} = require("addon/lib/utils");
 
 Cu.import("resource://gre/modules/Task.jsm");
+Cu.import("resource://gre/modules/NewTabUtils.jsm");
 
 module.exports = class TopSitesFeed extends Feed {
-  constructor(...args) {
-    super(...args);
+  constructor(options) {
+    super(options);
     this.getScreenshot = getScreenshot;
     this.missingData = false;
+    this.pinnedLinks = options.pinnedLinks || NewTabUtils.pinnedLinks;
   }
   shouldGetScreenshot(link) {
     const isMissingIcon = !link.favicon_width && !link.favicon_height;
     const hasSmallIcon = !isMissingIcon && ((link.favicon_width < MIN_ICON_SIZE) || (link.favicon_height < MIN_ICON_SIZE));
     const badIcon = link.hasMetadata && (isMissingIcon || hasSmallIcon);
     return !isRootDomain(link.url) || badIcon;
   }
+  pinLink(action) {
+    this.pinnedLinks.pin(action.data.site, action.data.index);
+  }
+  unpinLink(action) {
+    this.pinnedLinks.unpin(action.data.site);
+  }
+  sortLinks(links, pinned) {
+    // Seperate out the pinned from the rest
+    let pinnedLinks = links.filter(link => this.pinnedLinks.isPinned(link));
+    let sortedLinks = links.filter(link => !this.pinnedLinks.isPinned(link));
+
+    // Insert the pinned links in their location
+    pinned.forEach((val, index) => {
+      if (!val) { return; }
+      let site = Object.assign({}, pinnedLinks.find(link => link.url === val.url), {isPinned: true, pinIndex: index});
+      if (index > sortedLinks.length) {
+        sortedLinks[index] = site;
+      } else {
+        sortedLinks.splice(index, 0, site);
+      }
+    });
+
+    return sortedLinks;
+  }
   getData() {
     return Task.spawn(function*() {
       const experiments = this.store.getState().Experiments.values;
 
-      let links;
-      // Get links from places
-      links = yield PlacesProvider.links.getTopFrecentSites();
+      // Get pinned links
```

Comment:
Yeah, I think I'll hold off on adding the `excludedLinks` argument to the query since those queries are being commited to m-c. We should probably do the change over there once it lands.